### PR TITLE
Port environments.js to @flow.

### LINKED ISSUE #859: Port environments.js to @flow

### PR DESCRIPTION
**Pull request by marcianx**
Fixes #859.
Sits on top of [PR 860](https://github.com/Khan/KaTeX/pull/860), which needs to be submitted first!

### CODE REVIEW

File: src/environments.js [hunk 1]
```diff
@@ -1,18 +1,96 @@
...
+ *  - envName: the name of the environment, one of the listed names.
+ *  - parser: the parser object.
+ *  - positions: the positions associated with these arguments from args.
+ */
+type EnvContext = {
+  mode: Mode,
+  envName: string,
+  parser: Parser,
+  positions: number[],
+};
```
Comment (kevinbarabash) [inline src/environments.js]: nit: 4 space tab.

File: src/environments.js [hunk 2]
```diff
@@ -1,18 +1,96 @@
...
+
+/**
+ * List of ParseNodes followed by an array of positions.
+ * Returned by `Parser`'s `parseArguments`.
+ */
+type ParseResult = (ParseNode | number[])[] | ParseNode;
```
Comment (kevinbarabash) [inline src/environments.js]: When do we return a result with type `number[][]`?

File: src/environments.js [hunk 3]
```diff
@@ -1,18 +1,96 @@
...
+};
+
+type EnvData = {
+    numArgs: number,
+    argTypes?: ArgType[],
+    greediness: number,
+    allowedInText: boolean,
+    numOptionalArgs: number,
+    handler: EnvHandler,
+};
```
Comment (kevinbarabash) [inline src/environments.js]: This looks suspiciously similar to how functions are typed.  We might be able to dedupe in future.

File: src/environments.js [hunk 4]
```diff
@@ -1,18 +1,96 @@
...
+ *  - context: information and references provided by the parser
+ *  - args: an array of arguments passed to \begin{name}
+ */
+type EnvHandler = (context: EnvContext, args: ParseNode[]) => ParseResult;
+
+/**
+ *  - numArgs: (default 0) The number of arguments after the \begin{name} function.
+ *  - argTypes: (optional) Just like for a function
+ *  - allowedInText: (default false) Whether or not the environment is allowed
+ *  i                nside text mode (not enforced yet).
```
Comment (kevinbarabash) [inline src/environments.js]: `i nside => inside`

File: src/environments.js [hunk 5]
```diff
@@ -1,18 +1,96 @@
...
+export default environments;
+
+// Data stored in the ParseNode associated with the environment.
+type ArrayColSpec = { type: "separator", separator: string } | {
+    type: "align",
+    align: string,
+    pregap?: number,
+    postgap?: number,
+};
+type EnvNodeData = {
```
Comment (kevinbarabash) [inline src/environments.js]: `ArrayEnvNode` or `ArrayNode` would be better as this type is specific to arrays.  For consistency `ArrayColSpec` could be `AlignNode` and we could even introduce a `SeparateNode` type.

File: src/environments.js [hunk 6]
```diff
@@ -1,18 +1,96 @@
...
+function parseArray(
+    parser: Parser,
+    result: EnvNodeData,
+    style: StyleStr,
+) {
     let row = [];
     const body = [row];
     const rowGaps = [];
-    while (true) {
+    for (;;) {
```
Comment (kevinbarabash) [inline src/environments.js]: Why the change?

File: src/environments.js [hunk 7]
```diff
@@ -1,18 +1,96 @@
...
  * columns delimited by &, and create a nested list in row-major order
  * with one group per cell.  If given an optional argument style
  * ("text", "display", etc.), then each cell is cast into that style.
  */
-function parseArray(parser, result, style) {
+function parseArray(
+    parser: Parser,
+    result: EnvNodeData,
+    style: StyleStr,
+) {
```
Comment (kevinbarabash) [inline src/environments.js]: Thank you.

File: src/environments.js [hunk 8]
```diff
@@ -42,35 +120,24 @@ function parseArray(parser, result, style) {
...
- *  - envName: the name of the environment, one of the listed names.
- *  - parser: the parser object
- *  - lexer: the lexer object
- *  - positions: the positions associated with these arguments from args.
- * The handler must return a ParseResult.
+ * - props can either be an EnvProps struct, or a number indicating the numArgs.
  */
-function defineEnvironment(names, props, handler) {
+function defineEnvironment(
+    names: string | string[],
```
Comment (kevinbarabash) [inline src/environments.js]: We updated `defineFucntion` to only take an array of `strings`.  We should probably do the same here.

File: src/environments.js [hunk 9]
```diff
@@ -42,35 +120,24 @@ function parseArray(parser, result, style) {
...
+    props: EnvProps | number,
```
Comment (kevinbarabash) [inline src/environments.js]: `defineFunction` doesn't support taking a `number` for `props`, we should do the same here too.

File: src/environments.js (hunk 1 above)
Comment (marcianx) [reply src/environments.js]: Oi! Time to change my vim defaults. I keep forgetting to set it to 4 spaces in every new session. ;)

File: src/environments.js (hunk 2 above)
Comment (marcianx) [reply src/environments.js]: `Parser`'s `parseArgument` always returns the list of positions as the last element.
So it returns something of the format `[ParseNode, ParseNode, ..., ParseNode, number[]]`. So this happens any time you see a `defineFunction` with code like
```javascript
    res = parseArray(context.parser, res, "display");
    return res;
```
We need to refactor this at some point, but I'd rather do it right before or during `ParseNode` being ported. It would help to have maximal support from type-safety while doing this since `ParseResult` is simply stored here and used various places (including possible buildHtml, but I haven't read that part of the code base yet).

File: src/environments.js (hunk 6 above)
Comment (marcianx) [reply src/environments.js]: Ah, that was just to get rid of the `eslint` suppression at the top (personally a fan of maximal linting). However, that's not my favorite lint anyway; and I can revert this if you like.

File: src/environments.js (hunk 3 above)
Comment (marcianx) [reply src/environments.js]: Agreed.

File: src/environments.js (hunk 4 above)
Comment (marcianx) [reply src/environments.js]: Fixed.

File: src/environments.js (hunk 5 above)
Comment (marcianx) [reply src/environments.js]: `ArrayEnvNode` is more descriptive. So I'm using that.
Is `AlignNode` the most appropriate name? It's not a `ParseNode`.

Comment (marcianx) [reply src/environments.js]: Also, per `ArrayEnvNode`, it's not a `ParseNode`, but the data stored in the `ParseNode`. Would you still prefer to call it `ArrayEnvNode`? Or would `ArrayEnvNodeData` (a little long, I know) be more appropriate to avoid confusion?

Comment (marcianx) [reply src/environments.js]: In the currently-pushed version, I went with `AlignSpec` and `ArrayEnvNodeData`. I'm very open to changing per your feedback.

File: src/environments.js (hunk 8 above)
Comment (marcianx) [reply src/environments.js]: Done.

File: src/environments.js (hunk 9 above)
Comment (marcianx) [reply src/environments.js]: Done.

File: src/environments.js (hunk 6 above)
Comment (kevinbarabash) [reply src/environments.js]: Gotcha.  This is fine.  We can re-evaluate lint rules separately.

File: src/environments.js (hunk 2 above)
Comment (kevinbarabash) [reply src/environments.js]: I see, it's documented in the comment above.

> We need to refactor this at some point, but I'd rather do it right before or during ParseNode being ported.

Sounds good.

File: src/environments.js (hunk 5 above)
Comment (kevinbarabash) [reply src/environments.js]: I don't see a need to differentiate parse nodes from environment nodes as they end up being part of the same tree at the end of the day.  I think eventually we'll types for each of the different nodes the parser can generate and eventually `ParseNode` will become:
```
ParseNode = FracNode | ExpNode | ... | ArrayNode | AlignNode | ... 
```
Environments are just one type of thing that the parser parses as it does its work, it just happens that the environment parsing code is in a separate file.

Comment (marcianx) [reply src/environments.js]: I'm sorry, I didn't quite understand this comment and the future evolution you intend. What I meant was: for `result: ArrayEnvNodeData`, we have
```javascript
    return new ParseNode(result.type, result, parser.mode);
```
So the `ParseNode` instance contains a _payload_ that is of type `ArrayEnvNodeData` struct, which is why I was thinking about the `Data` suffix.

So did you mean that the payload would become a union type (e.g. `ParseNodeData` or something briefer)?
I'm guessing you didn't mean that the `ParseNode` class would be dissolved in favor of a union type (given that we still want to encapsulate the common tokenizer metadata for error reporting)?

File: src/environments.js (hunk 3 above)
Comment (marcianx) [reply src/environments.js]: FYI: Looking at `defineFunction.js` I don't currently see any benefit in coalescing.
1. We don't really need any subtype relationship.
2. For object/struct types, there's not really an `extends` type behavior AFAICS (please correct me if I'm wrong), so we would need to repeat the fields anyway when defining it.

File: src/environments.js (hunk 5 above)
Comment (kevinbarabash) [reply src/environments.js]: I was thinking that eventually we'd have a union type to specify the different types in the tree.  It's possible that `ParseNode` wraps these data nodes and adds additional data like error reporting.  I have spent any time thinking trough the details.  The high-level goal is for nodes with in the parse tree to be more regular and for all them to be statically typed.

Comment (marcianx) [reply src/environments.js]: Ok, so what you're saying is that the _payload_ of `ParseNode` is still considered an undecorated "Node". And hence we should use the `Node` suffix rather than `NodeData`?
Is that your preference?

**Comment by kevinbarabash** [comment]
@marcianx thanks for the PR.  It's exciting to see all of these implicit types being made explicit.

File: src/environments.js (hunk 5 above)
Comment (marcianx) [reply src/environments.js]: Ah, I see that you merged it. Please LMK what nomenclature you had in mind. I'm happy to update these as we continue in the future.
